File: CoUML-app/CoUML TS/Relationship.ts
import { User } from "./User";
import { RootObject } from "./RootObject";
import { Component } from "./Component";
import { AttributeCollection } from "./Attribute";

enum RelationshipType{
    Dependency,
    Association,
    Aggregation,
    Composistion,
    Generalization
}

export class Relationship implements RootObject{
    editor:User;
    type:RelationshipType;
    fromComponent:Component;
    toComponent:Component;
    atributes:AttributeCollection;
}

export interface RelationshipCollectionIF{
    Iterator():RelationshipIteratorIF;
}

//after that rewrite in c# for both ends
//collections are part of iterator pattern
export class RelationshipCollection implements RelationshipCollectionIF{
    Relationships: Relationship[];
    
    //creates an iteratorIF
    Iterator():RelationshipIteratorIF{
        return new RelationshipIterator();
    }
}

export interface RelationshipIteratorIF{
    hasNextRelationship():boolean;
    getNextRelationship():Relationship;
    hasPreviousRelationship():boolean;
    getPreviousRelationship():Relationship;
}

//fetches objects from collection
export class RelationshipIterator implements RelationshipIteratorIF{
    collection:RelationshipCollection;
    position:number = 0;

    hasNextRelationship():boolean{
        return this.collection.Relationships.length > this.position+1;
    }
    getNextRelationship():Relationship{
        if(this.hasNextRelationship){
            return this.collection.Relationships[++this.position];
        }
    }
    hasPreviousRelationship():boolean{
        return this.position-1 > 0;
    }
    getPreviousRelationship():Relationship{
        if(this.hasPreviousRelationship){
            return this.collection.Relationships[--this.position];
        }
    }
}

File: CoUML-app/Models/Relationship.cs
/*
using User;
using RootObject;
using Component;
*/

enum RelationshipType{
    Dependency,
    Association,
    Aggregation,
    Composistion,
    Generalization
}

class Relationship : RootObject{
    User editor;
    RelationshipType type;
    Component fromComponent;
    Component toComponent;
    AttributeCollection attributes;
}

interface RelationshipCollectionIF{
    RelationshipIteratorIF Iterator();
}

class RelationshipCollection : RelationshipCollectionIF{
    public Relationship[] relationships;
    RelationshipIteratorIF RelationshipCollectionIF.Iterator(){
        return new RelationshipIterator();
    }
}

interface RelationshipIteratorIF{
    bool hasNextRelationship();
    Relationship getNextRelationship();
    bool hasPreviousRelationship();
    Relationship getPreviousRelationship();
}

class RelationshipIterator : RelationshipIteratorIF{
    RelationshipCollection collection;
    int position = 0;

    bool RelationshipIteratorIF.hasNextRelationship(){
        return this.collection.relationships.Length > this.position+1;
    }
    Relationship RelationshipIteratorIF.getNextRelationship(){
        RelationshipIteratorIF iterator = this;
        if (iterator.hasNextRelationship()){
            return this.collection.relationships[++this.position];
        }
        return null;
    }
    bool RelationshipIteratorIF.hasPreviousRelationship(){
        return this.position-1 > 0;
    }
    Relationship RelationshipIteratorIF.getPreviousRelationship(){
        RelationshipIteratorIF iterator = this;
        if(iterator.hasPreviousRelationship()){
            return this.collection.relationships[--this.position];
        }
        return null;
    }
}